File: backend/app/schemas/websocket.py
import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel


class WebSocketEventType(str, Enum):
    MESSAGE = "message"
    TYPING = "typing"
    STOP_TYPING = "stop_typing"
    READ = "read"
    USER_CONNECTED = "user_connected"
    USER_DISCONNECTED = "user_disconnected"


class BaseWebSocketEvent(BaseModel):
    type: WebSocketEventType


class MessageEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.MESSAGE
    content: str


class TypingEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.TYPING


class StopTypingEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.STOP_TYPING


class ReadEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.READ
    message_id: uuid.UUID


class OutgoingMessageEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.MESSAGE
    id: uuid.UUID
    conversation_id: uuid.UUID
    sender_id: uuid.UUID
    sender_role: str
    content: str
    sent_at: datetime


class UserConnectedEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.USER_CONNECTED
    user_id: uuid.UUID


class UserDisconnectedEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.USER_DISCONNECTED
    user_id: uuid.UUID


class TypingResponseEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.TYPING
    user_id: uuid.UUID


class StopTypingResponseEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.STOP_TYPING
    user_id: uuid.UUID


class ReadResponseEvent(BaseWebSocketEvent):
    type: WebSocketEventType = WebSocketEventType.READ
    message_id: uuid.UUID
    read_at: datetime

File: frontend/src/services/Chats/websocket.ts
import { WS_URL } from '../../config/api';
import { tokenStorage } from '../../utils/tokenStorage';

export interface SocketMessage {
  type: 'message' | 'typing' | 'stop_typing' | 'read' | 'user_connected' | 'user_disconnected';
  user_id?: string;

  id?: string;
  conversation_id?: string;
  sender_id?: string;
  sender_role?: 'patient' | 'nutritionist' | 'assistant';
  content?: string;
  sent_at?: string;
  read_at?: string;
}

export class ChatSocket {
  private socket?: WebSocket;

  connect(
    conversationId: string,
    onMessage: (data: SocketMessage) => void,
    onOpen?: () => void,
    onClose?: () => void,
    onError?: (event: Event) => void,
  ) {
    const token = tokenStorage.get();

    if (!token) {
      throw new Error('No existe token de autenticación');
    }

    this.socket = new WebSocket(`${WS_URL}/ws/${conversationId}?token=${token}`);

    this.socket.onopen = () => {
      onOpen?.();
    };

    this.socket.onmessage = (event: MessageEvent) => {
      try {
        const data: SocketMessage = JSON.parse(event.data);
        onMessage(data);
      } catch (err) {
        console.error('Error parseando mensaje', err);
      }
    };

    this.socket.onerror = (event: Event) => {
      console.error('WebSocket error');
      onError?.(event);
    };

    this.socket.onclose = () => {
      console.log('WebSocket desconectado');
      onClose?.();
    };
  }

  sendMessage(content: string) {
    this.send({
      type: 'message',
      content,
    });
  }

  sendTyping() {
    this.send({
      type: 'typing',
    });
  }

  sendStopTyping() {
    this.send({
      type: 'stop_typing',
    });
  }

  sendRead() {
    this.send({
      type: 'read',
    });
  }

  private send(data: object) {
    if (!this.socket) return;

    if (this.socket.readyState !== WebSocket.OPEN) {
      console.warn('WebSocket aún no está conectado');
      return;
    }

    this.socket.send(JSON.stringify(data));
  }

  disconnect() {
    this.socket?.close();
    this.socket = undefined;
  }

  isConnected(): boolean {
    return this.socket?.readyState === WebSocket.OPEN;
  }
}
